src/datadog_directive.cpp

```diff
@@ -797,5 +798,144 @@ char *set_datadog_agent_url(ngx_conf_t *cf, ngx_command_t *command, void *conf)
       });
 }
 
+char *set_datadog_delegate_sampling(ngx_conf_t *cf, ngx_command_t *command, void *conf) noexcept {
+  const auto loc_conf = static_cast<datadog_loc_conf_t *>(conf);
+  loc_conf->sampling_delegation_directive = command_source_location(command, cf);
+
+  auto values = static_cast<ngx_str_t *>(cf->args->elts);
+  // values[0] is the command name, "datadog_delegate_sampling".
+  // The other elements are the arguments: either zero or one of them.
+  //
+  //     datadog_delegate_sampling [on | off];
+
+  const int num_args = cf->args->nelts - 1;
+
+  ngx_str_t pattern;
+  if (num_args == 0) {
+    pattern = ngx_string("on");
+  } else {
+    assert(num_args == 1);
+    pattern = values[1];
+  }
```

Comment:
I like that `datadog_delegate_sampling;` delegates the sampling. Just what's written on the tin, as they say.

I was about to say that it's consistent with `datadog_trace_locations;` and `datadog_trust_incoming_span;` but based on the definition of those directives (both inherited from nginx-opentracing), it looks like they require the argument.

I still prefer the way I have `datadog_delegate_sampling`.

The only drawback that I noticed is that there are now two notions of what is meant by "default." There's the default if you don't mention the directive, and then there's the default if you mention the directive but don't provide an argument. And those defaults are different.

I don't feel too strongly about this, but I think `datadog_delegate_sampling;` is a good shorthand. The reason is that the only two real use cases are "don't mention it" and "mention it so it's on." The `on|off` argument, and its ability to depend on variable expressions, is just for generality.

Reply:
I was about to change my mind about this, but then looking at it again I _still_ prefer to keep `datadog_delegate_sampling;` as being a shorthand for `datadog_delegate_sampling on;`.

Can you do more to convince me? There's a consistency argument to be made. Here are the `on|off` directives:

- `datadog_sample_rate <rate> [on|off]`
    - `on|off` can be a variable expression
    - if `on|off` is omitted, it defaults to `on`
- `datadog_delegate_sampling [on|off]`
    - `on|off` can be a variable expression
    - if `on|off` is omitted, it defaults to `on`
- `datadog_trust_incoming_span on|off`
    - `on|off` is required, and cannot be a variable expression
- `datadog_trace_locations on|off`
    - `on|off` is required, and cannot be a variable expression
- `datadog_allow_sampling_delegation_in_subrequests [on|off]`
    - `on|off` can be a variable expression
    - if `on|off` is omitted, it defaults to `on`

So, there are two styles at play:

1. `on|off` is an optional variable expression that defaults to `on`
2. `on|off` is a required literal string

Style (2) is inherited from [nginx-opentracing][1]. Style (1) is how I do.

We could convert everything to style (1), and it would be backward compatible.

[1]: https://github.com/opentracing-contrib/nginx-opentracing

Reply:
> Can you do more to convince me?

- As you mentioned it is inconsistent.
- `datadog_delegate_sampling on;` expresses intent better than `datadog_delegate_sampling;`. Now I have to check what's the default behavior.
- The code handling those directives will be simpler:
````cpp
if (num_args < 1) {
   return; // nope
}
ngx_str_t pattern = values[1];

// instead of
 
ngx_str_t pattern;
if (num_args == 0) {
    pattern = ngx_string("on");
} else {
    assert(num_args == 1);
    pattern = values[1];
}
````

Do what please you.

Reply:
Alright, thanks. I'll give it another think.